### ISSUE TITLE
ACME RFC2136 update not retried on error

### ISSUE THREAD

**Issue by pdf** [containous/traefik#4344]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

Set up RFC2136 acme certificate updates.

### What did you expect to see?

Certs updated even in the case of transient errors with the configured DNS server, or network issues like packet loss.

### What did you see instead?

- At renewal, either the server specified by RFC2136_NAMESERVER or the network appears to have had issues, resulting in the following:
  `msg="Error renewing certificate from LE: {*.mydomain.com [mydomain.com]}, acme: Error -> One or more domains had a problem:\n[mydomain.com] error presenting token: rfc2136: DNS update failed: read udp [2001:xxxx::1]:52529->[2001:xxxx::2]:53: i/o timeout\n"`
- Subsequently, the following message was repeated every day, so it appears to be trying to confirm the DNS update that never happened, and failing:
  `msg="Error renewing certificate from LE: {*.mydomain.com [mydomain.com]}, acme: Error -> One or more domains had a problem:\n[mydomain.com] time limit exceeded: last error: NS ns0.mydomain.com. did not return the expected TXT record [fqdn: _acme-challenge.mydomain.com.]\n"`

Now, my certs are going to expire in a couple of weeks, and I don't know how to make traefik retry the DNS update so that the renewal can occur.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.7.6
Codename:     maroilles
Go version:   go1.11.3
Built:        2018-12-14_06:43:37AM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
[acme]
email = "letsencrypt@mydomain.com"
storage = "/etc/traefik/acme/acme.json"
entryPoint = "https"

# staging server for testing
#caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
acmeLogging = true

        [acme.httpChallenge]
        entryPoint = "http"

        [acme.dnsChallenge]
        provider = "rfc2136"

        [[acme.domains]]
        main = "*.mydomain.com"
        sans = ["mydomain.com"]

        [[acme.domains]]
        main = "*.myotherdomain.com"
        sans = ["myotherdomain.com"]
```

**Comment by jlevesy** [comment]
Hi @pdf, thanks for your interest in the project.

I'm digging into that right now, I will get back to you soon.

However, I have to say that you have currently both the HTTP challenge and the DNS challenge enabled, so the DNS challenge will prevail as [shown here](https://github.com/containous/traefik/blob/v1.7/configuration/configuration.go#L393).

I don't know your use case, but maybe you can disable the `dnsChallenge` and fallback to the `httpChallenge` and it will get you out of trouble temporarily.

**Comment by pdf** [comment]
Hi @jlevesy I'm using wildcards, and have a lot of subdomains enabled, so I don't think HTTP challenge is an option - I tried it against staging when I originally set this up, and must have missed removing that stanza, but moved to DNS to avoid the massive cert request flood.

**Comment by ldez** [comment]
You got the timeout because your DNS don't answer after 60s: during this time, every 2s, Traefik retry to call your DNS.

You can change timeout and interval with this env vars:
- `RFC2136_TIMEOUT` (default 60) in seconds
- `RFC2136_POLLING_INTERVAL` (default: 2) in seconds

**Comment by pdf** [comment]
Right, but what can I do about it now that it's failed? That's the problem.

**Comment by pdf** [comment]
@ldez the timeout has already occurred though.  Are you telling me that if I set the timeout to 2 weeks or something, that traefik will start trying again?

**Comment by ldez** [comment]
Not at all.

You receive the error every day because you are in the renewal period (30 before expiration), so Traefik tries to renew it every day, but every attempt fails.

So you have increase the timeout to several minutes (not weeks or day) to allow Traefik to renew your certificate.

**Comment by pdf** [comment]
Okay, I think I understand you now - the logging does not include successful operations, and only produces a single log message in case of failure.  This is likely part of the source of my confusion - the earlier log message about a write error is in fact irrelevant, and the timeout logs each day relate to a complete round-trip attempt at renewal.

How are the TXT records resolved to verify that the update operation succeeded? Are recursors queried, or the RFC2136_NAMESERVER, or only the public authoritative server(s) for the domain? This will determine the required timeout for propagation.

**Comment by ldez** [comment]
Your error is related to `RFC2136_NAMESERVER` answer.

The error is related to the addition of the TXT record in your DNS, and it's not exactly related to the propagation, but it's the same timeout.

**Comment by ldez** [comment]
The question has been answered.

Join our [Slack workspace](https://slack.traefik.io/) for more [community #support](https://traefik.slack.com/messages/C0CDT22PJ/).

**Comment by pdf** [comment]
Aside from the fact that some better logging would go a long way to improving the ability of users to understand what's happening, I'm not sure this is operating as described - I adjusted `RFC2136_TIMEOUT`  to `600s` and restarted traefik.  This resulted in a failed update attempt:

```
time="2019-01-04T00:50:00+11:00" level=error msg="Error renewing certificate from LE: {*.mydomain.com [mydomain.com]}, acme: Error -> One or more domains had a problem:\n[mydomain.com] time limit exceeded: last error: NS ns1.mydomain.com. did not return the expected TXT record [fqdn: _acme-challenge.mydomain.com.]\n"
```

But compare the DNS server log (times are sync'd from the same NTP source on these servers):

```
Jan 04 00:48:59 nsserver.mydomain.com named-pkcs11[1079]: client 2100:xxxx::1#52881/key traefik: updating zone mydomain.com/IN: deleting rrset at _acme-challenge.mydomain.com TXT
Jan 04 00:48:59 nsserver.mydomain.com named-pkcs11[1079]: client 2100:xxxx::1#52881/key traefik: updating zone mydomain.com/IN: adding an RR at _acme-challenge.mydomain.com TXT "lime6QRf5GDyjyoA1J6Y3ckHI4pFoS1gx72cBatA6nZ"
Jan 04 00:48:59 nsserver.mydomain.com named-pkcs11[1079]: client 2100:xxxx::1#46461/key traefik: updating zone mydomain.com/IN: deleting rrset at _acme-challenge.mydomain.com TXT
Jan 04 00:48:59 nsserver.mydomain.com named-pkcs11[1079]: client 2100:xxxx::1#46461/key traefik: updating zone mydomain.com/IN: adding an RR at _acme-challenge.mydomain.com TXT "THozf4OMzsBccRLo6Ql3bjcGLb4z54blC5qyNRatSVw"
Jan 04 00:48:59 nsserver.mydomain.com named-pkcs11[1079]: zone mydomain.com/IN: sending notifies (serial 1546523343)
Jan 04 00:49:00 nsserver.mydomain.com named-pkcs11[1079]: client 10.10.10.10#57477 (mydomain.com): transfer of mydomain.com/IN: IXFR started (serial 1546520675 -> 1546523343)
Jan 04 00:49:00 nsserver.mydomain.com named-pkcs11[1079]: client 10.10.10.10#57477 (mydomain.com): transfer of mydomain.com/IN: IXFR ended
```

You can see number of things from these logs:
1. The update was performed, and in fact even the slave was notified and completed the IXFR within 1 second.
2. The `RFC2136_TIMEOUT` is not respected - the timeout log from traefik has a timestamp of 00:50:00, which would be 60s, not 600s.
3. Even though the update was performed (and I can confirm that the updated value from the log is the same value served by all servers in the cluster), and performed well within the timeout period, traefik still believes the update failed.

I'm also curious why there are two updates to the challenge record in quick succession with different values, and whether this may be related to the issue. This behaviour is repeated for my other domains.

**Comment by ldez** [comment]
I confirm the first record erases the second one.
It's a bug, and I already made a fix about that https://github.com/xenolf/lego/pull/745

**Comment by pdf** [comment]
I see this has been merged upstream, are there any docs on how to produce a custom build of traefik? I'm running out of time on my certs.

**Comment by ldez** [comment]
I will create a PR on Traefik soon.

**Comment by traefiker** [comment]
Closed by #4376.
